### PR TITLE
Significant updates to nftables support

### PR DESCRIPTION
**Pull request by paul-chambers**
Branch with significant changes that updates nftables support in miniupnpd to work with current versions of libnftnl (and nftables underneath). Also includes some maintainability and readability improvements.

Also tested independently by @4np. I believe this fixes issue #397

### CODE REVIEW

**Comment by paul-chambers** [comment]
Build appears to be failing because `config.h` isn't being regenerated with `USE_NFTABLES `set. Builds without error for me.

File: miniupnpd/netfilter_nft/nftnlrdr.c [hunk 1]
```diff
@@ -219,11 +325,11 @@ delete_filter_rule(const char * ifname, unsigned short port, int proto)
 	struct nftnl_rule *r;
 	UNUSED(ifname);
 
-	reflesh_nft_cache_filter();
+	refresh_nft_cache_filter();
```
Comment (miniupnp) [inline miniupnpd/netfilter_nft/nftnlrdr.c]: what is changed ?

Comment (miniupnp) [reply miniupnpd/netfilter_nft/nftnlrdr.c]: ah ok it was a typo ;)

File: miniupnpd/netfilter_nft/nftnlrdr_misc.c [hunk 2]
```diff
@@ -32,23 +32,45 @@
...
+#include <libnftnl/chain.h>
 #include <libnftnl/rule.h>
 #include <libnftnl/expr.h>
 
+#include "../commonrdr.h"
 #include "nftnlrdr_misc.h"
 #include "../macros.h"
-#include "../upnpglobalvars.h"
+
+#define DEBUG 1
```
Comment (miniupnp) [inline miniupnpd/netfilter_nft/nftnlrdr_misc.c]: that's an issue. You should build with` CFLAGS=-DDEBUG` or `#define DEBUG 1` in config.h if you want to enable debug

File: miniupnpd/commonrdr.h [hunk 3]
```diff
@@ -64,4 +64,20 @@ update_portmapping_desc_timestamp(const char * ifname,
...
+/* only provided by nftables implementation at the moment */
+
+typedef enum {
+	TABLE_NAME,
+	TABLE4_NAME,
+	TABLE6_NAME,
+	NAT_CHAIN_NAME,
+	NAT_POSTROUTING_CHAIN_NAME,
+	FORWARD_CHAIN_NAME,
+} rdr_name_type;
```
Comment (miniupnp) [inline miniupnpd/commonrdr.h]: shouldn't you use more specific names like NF_TABLE_NAME ?

File: miniupnpd/miniupnpd.c [hunk 4]
```diff
@@ -1267,6 +1267,17 @@ init(int argc, char * * argv, struct runtime_vars * v)
...
+			case UPNPFORWARDCHAIN:
+				set_rdr_name(FORWARD_CHAIN_NAME, ary_options[i].value);
+				break;
+			case UPNPNATCHAIN:
+				set_rdr_name(NAT_CHAIN_NAME, ary_options[i].value);
+				break;
+			case UPNPNATPOSTCHAIN:
+				set_rdr_name(NAT_POSTROUTING_CHAIN_NAME, ary_options[i].value);
+				break;
+#else
```
Comment (miniupnp) [inline miniupnpd/miniupnpd.c]: the same modification can apply to standard iptables, right ?
I think miniupnpd_forward_chain/miniupnpd_nat_chain/miniupnpd_nat_postrouting_chain are only used in iptcrdr.c

File: miniupnpd/miniupnpd.c [hunk 5]
```diff
@@ -2901,6 +2913,8 @@ main(int argc, char * * argv)
 	freeoptions();
 #endif
 
+	shutdown_redirect();
```
Comment (miniupnp) [inline miniupnpd/miniupnpd.c]: nice one !

Comment (miniupnp) [reply miniupnpd/miniupnpd.c]: thanks for spoting that !

**Comment by miniupnp** [comment]
> CI build appears to be failing because `config.h` isn't being regenerated with `USE_NFTABLES `set. Builds without error for me.

that's fixed ;)

File: miniupnpd/netfilter_nft/nftnlrdr_misc.c (hunk 2 above)
Comment (paul-chambers) [reply miniupnpd/netfilter_nft/nftnlrdr_misc.c]: Oops... sorry, that was meant to be a temporary thing, I just forgot to remove it. I've noticed I have a couple of debug messages I also meant to remove. I'll commit a fix.

File: miniupnpd/miniupnpd.c (hunk 4 above)
Comment (paul-chambers) [reply miniupnpd/miniupnpd.c]: Yes, that's the intent. I just haven't been able to test the IPtables version, so I limited what I changed to what I could test easily. More a matter of time than anything else.

The way nat works on nftables, you must have both the prerouting and postrouting chains defined (to establish both inbound and outbound paths) irrespective of whether both chains have rules. So even though no rules are added to 'postrouting', it must be present or the rules in 'prerouting' won't work properly.
 
(part of my nefarious plan to implement privilege separation ;)

File: miniupnpd/commonrdr.h (hunk 3 above)
Comment (paul-chambers) [reply miniupnpd/commonrdr.h]: Good point - enums are scoped to the class that contains them in C++, but they have file scope in C - I'd forgotten that. I'll make them more unique.

File: miniupnpd/miniupnpd.c (hunk 5 above)
Comment (paul-chambers) [reply miniupnpd/miniupnpd.c]: Didn't really matter while shutdown_redirect() was a stub. So it was a problem of my own creation :)

**Comment by paul-chambers** [comment]
Thanks for the Travis fix. I think I've resolved everything.

I caught another bug in the original code. The 'description' is stored in USERDATA, a sized blob. Thus you can't just use strdup() to copy it as if it's a null-terminated string, must use strndup().

**Comment by miniupnp** [comment]
merged manually

**Pull request closed without merging**